### ISSUE TITLE
Failure to tag security group prevents cluster reconciliation

### ISSUE THREAD

**Issue by sethp-nr** [kubernetes-sigs/cluster-api-provider-aws#890]
/kind bug

**What steps did you take and what happened:**

1. Followed our usual process for pivoting a new cluster
2. Part way through, the EC2 API's eventual consistency whacked us with a 404 trying to tag the security group CAPA had just created (see log message below)
3. From then on out, CAPA couldn't find the security group, so it tried to create it again, getting a conflict error on the SG's "group name" (which is different from the Name tag, because AWS). That error kept us from getting past the `reconcileNetwork` step to creating the load balancer & everything later

It's sort of similar to #851, but with different consequences because security groups have a name-based identity.

**What did you expect to happen:**

I think for security groups, we'll have a better time if we don't filter by cluster tag here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/46d5f38/pkg/cloud/aws/services/ec2/securitygroups.go#L165

That said, when we added that tag by hand, we then saw a different error trying to tag the resource:

```
E0711 14:17:57.002589       1 controller.go:134] Error reconciling cluster object gallant-agnesi; failed to reconcile network for cluster "gallant-agnesi": failed to ensure tags on security group "sg-SGIDf": failed to tag resource "" in cluster "gallant-agnesi": InvalidParameterValue: Value ( null ) for parameter resourceId is invalid. Null/empty value for resourceId is invalid
```

**Anything else you would like to add:**

Some log messages (I found "the security group [extremely specific ID] does not exist" pretty funny):

```
E0711 14:08:11.908350       1 controller.go:134] Error reconciling cluster object gallant-agnesi; failed to reconcile network for cluster "gallant-agnesi": failed to tag security group "gallant-agnesi-lb" in vpc "vpc-VPCID": InvalidGroup.NotFound: The security group 'sg-SGID' does not exist
	status code: 400, request id: abd9cce5-5aa6-439c-8953-040b09e3f86c
...
E0711 14:13:53.275716       1 controller.go:134] Error reconciling cluster object gallant-agnesi; failed to reconcile network for cluster "gallant-agnesi": failed to create security group "gallant-agnesi-lb" in vpc "vpc-VPCID": InvalidGroup.Duplicate: The security group 'gallant-agnesi-lb' already exists for VPC 'vpc-VPCID'
```

**Environment:**

- Cluster-api-provider-aws version: issue looks the same on master

**Comment by sethp-nr** [comment]
I think there's a quick fix and a harder fix path here. The quick fix is (maybe) to yank the tag filter from the SG search.

The harder path would be a combination of retries on the initial tag call, and some way to heal from this situation, like parsing the "this resource already exists" error out of the create security group response.

**Comment by detiber** [comment]
@sethp-nr I hate to say it, but considering the eventual consistency, we probably need to implement retries with backoffs in quite a few different places anyway.

**Comment by rudoi** [comment]
I think the harder path is probably the better one. :\ I think having the system attempt to heal itself in any way it can with respect to the EC2 APIs is a great direction to move in.

**Comment by rudoi** [comment]
Also I believe related to #836

**Comment by sethp-nr** [comment]
@detiber Is there a "I don't enjoy giving this a +1, but begrudgingly accept its necessity" emoji?

**Comment by detiber** [comment]
@sethp-nr not that I'm aware of, but I share the same sentiment. It'd be great if they built handling of this into the SDKs.

**Comment by ncdc** [comment]
Fixed by #916
/close
